extension/src/lib.rs

```diff
@@ -0,0 +1,295 @@
+        window_size: Milliseconds, // the size of a window to delta over
+        time: TimestampTz,
+        val: f64,
+    ) -> Option<Pox<GapfillDeltaTransition>> {
+        if time < lowest_time || time > greatest_time {
+            elog!(Error, "input time less than lowest time")
+        }
+
+        let mut state = state.unwrap_or_else(|| {
+            let state: Pox<_> = GapfillDeltaTransition::new(
+                lowest_time,
+                greatest_time,
+                window_size,
+                step_size,
+                false,
+                false,
+            ).into();
+            state
+        });
+
+        state.add_data_point(time, val);
+
+        Some(state)
+    }
+
+    pub fn gapfill_rate_transition(
+        state: Option<Pox<GapfillDeltaTransition>>,
+        lowest_time: TimestampTz,
+        greatest_time: TimestampTz,
+        step_size: Milliseconds, // `prev_now - step` is where the next window starts
+        window_size: Milliseconds, // the size of a window to delta over
+        time: TimestampTz,
+        val: f64,
+    ) -> Option<Pox<GapfillDeltaTransition>> {
+        if time < lowest_time || time > greatest_time {
+            elog!(Error, "input time less than lowest time")
+        }
+
+        let mut state = state.unwrap_or_else(|| {
+            let state: Pox<_> = GapfillDeltaTransition::new(
+                lowest_time,
+                greatest_time,
+                window_size,
+                step_size,
+                true,
+                true,
+            ).into();
+            state
+        });
+
+        state.add_data_point(time, val);
+
+        Some(state)
+    }
+
+    pub fn gapfill_increase_transition(
+        state: Option<Pox<GapfillDeltaTransition>>,
+        lowest_time: TimestampTz,
+        greatest_time: TimestampTz,
+        step_size: Milliseconds, // `prev_now - step` is where the next window starts
+        window_size: Milliseconds, // the size of a window to delta over
+        time: TimestampTz,
+        val: f64,
+    ) -> Option<Pox<GapfillDeltaTransition>> {
+        if time < lowest_time || time > greatest_time {
+            elog!(Error, "input time less than lowest time")
+        }
+
+        let mut state = state.unwrap_or_else(|| {
+            let state: Pox<_> = GapfillDeltaTransition::new(
+                lowest_time,
+                greatest_time,
+                window_size,
+                step_size,
+                true,
+                false,
+            ).into();
+            state
+        });
+
+        state.add_data_point(time, val);
+
+        Some(state)
+    }
+
+    pub fn gapfill_delta_final(state: Option<Pox<GapfillDeltaTransition>>) -> Option<*mut ArrayType> {
+        state.map(|mut s| s.to_pg_array())
+    }
+}
+
+#[derive(Debug)]
+struct GapfillDeltaTransition {
+    window: VecDeque<(TimestampTz, f64)>,
+    deltas: Vec<Datum>,
```

Comment:
lets leave comment that the delta and null vec are not same size. Rather the delta vec is of size = # items not null.